### ISSUE TITLE
Basic AACP framing info

### ISSUE THREAD

**Issue by vulpes2** [steam3d/MagicPodsCore#14]
Apple has renamed the protocol to AACP, still don't know what it stands for, maybe Apple Audio Configuration Protocol? Either way, here's the basic framing info for the AACP messages.

An AACP packet has a u16 type field, a u16 service ID field, and a payload. The service ID is always 4 after the connection has been established, other service IDs have not been observed so far.

There are five packet types:
- 0: Connect, payload: `u16 major, u16 minor, u64 features`
- 1: Connection Response, payload: `u16 status, u16 major, u16 minor, u64 features`
- 2: Disconnect, payload: `u16 status`
- 3: Disconnection Response, no payload
- 4: Message, payload: `u16 opcode, payload`

Payload for the AACP message packet varies depends on the opcode, they are either fixed size or will have a size field somewhere. For example, opcode 0x9 is a configuration message with a fixed payload of 5 bytes. The first byte is the configuration type, and the following 4 bytes are the values. Not all settings use all the 4 bytes, most will leave the unused bytes as 0x00.

More docs will follow in a few weeks, I just don't have time to write them yet. Handoff seems to be iCloud related and probably beyond my capabilities to figure out, most of my efforts so far have been centered around parsing all the AACP messages I can capture.

**Comment by steam3d** [comment]
AACP behaves differently depending on which capabilities are supported by the AirPods or Beats. Some Beats headphones do not have an infrared sensor; I noticed that on such headphones you can't control ANC. I suppose the iPhone sends different initialization packets, but the handshake message remains the same.

The interesting thing AirPods and Beats headphones does not notify some own settings intead of it the iPhone override settings during L2CAP connection and when you connect AirPods to iPad, the iPad will also override settings inside AirPods.

You can easily check this. Set up different settings for the 'Press and Hold' feature on the iPhone and iPad, and connect an Android phone after the iPhone. Try to switch between ANC modes and repeat the same with the iPad. You will notice that the settings are always overridden

**Comment by vulpes2** [comment]
That's consistent with the behavior I've seen so far. When connecting with my python script, the AirPods reports only a subset of the current settings, not sure if there is a way to dump all the current configuration. There also seems to be some certificate exchange stuff, which is probably for checking fake AirPods. The AirPods Pro seems to report its SN after the connection has been established without any special requests, but the original AirPods does not.

**Comment by steam3d** [comment]
Some fake AirPods also send firmware info and serial numbers. It depends on how good the replica is

**Comment by kavishdevar** [comment]
Find My (playing sound through speakers), and nearby stuff, like battery, device name, serial number, and a few more things (apparently some notifications too) use ATT (L2CAP with PSM 31) when it is not connected to the device! It is slightly weird that an iPad can connect to this socket without the airpods disconnecting the other device and stop playing audio (maybe the bluetooth stack on apple devices allows for socket-only connections, just like being able to control airpods settings from multiple macs at the same time idk)

But, I can't get it to work from linux. Probably some handshake packet is being sent, and of course I can not find that on newer versions of OS's as they're hidden

**Comment by steam3d** [comment]
@kavishdevar L2CAP classic or L2CAP LE?

**Comment by kavishdevar** [comment]
I'm not sure, how do I check? On linux, I connect using the same code as the AACP, just with a different PSM.. I can try using android's API which are separated for both types

**Comment by steam3d** [comment]
@kavishdevar Okay, I will check the number of services available on AirPods. The last time, I found only the AAP service.

**Comment by kavishdevar** [comment]
```
╭─kavish at BigBox in ~
╰─○ sudo sdptool browse 28:2D:7F:C2:05:5B
Browsing 28:2D:7F:C2:05:5B ...
Service Description: PnP Information
Service RecHandle: 0x4f491200
Service Class ID List:
  "PnP Information" (0x1200)
Language Base Attr List:
  code_ISO639: 0x656e
  encoding:    0x6a
  base_offset: 0x100
  code_ISO639: 0x6672
  encoding:    0x6a
  base_offset: 0x110
  code_ISO639: 0x6465
  encoding:    0x6a
  base_offset: 0x120
  code_ISO639: 0x6a61
  encoding:    0x6a
  base_offset: 0x130

Failed to connect to SDP server on 28:2D:7F:C2:05:5B: Connection refused
Service Name: GATT
Service RecHandle: 0x4f491801
Service Class ID List:
  "Generic Attribute" (0x1801)
Protocol Descriptor List:
  "L2CAP" (0x0100)
    PSM: 31 <--------- This one
  "ATT" (0x0007)
    uint16: 0x0066
    uint16: 0x0066
Language Base Attr List:
  code_ISO639: 0x656e
  encoding:    0x6a
  base_offset: 0x100
  code_ISO639: 0x6672
  encoding:    0x6a
  base_offset: 0x110
  code_ISO639: 0x6465
  encoding:    0x6a
  base_offset: 0x120
  code_ISO639: 0x6a61
  encoding:    0x6a
  base_offset: 0x130

Service Name: UARPS
Service RecHandle: 0x4f493878
Service Class ID List:
  UUID 128: 4715650b-5e9d-4ac2-b898-a4fc0aa5df78
Protocol Descriptor List:
  "L2CAP" (0x0100)
    PSM: 31
  "ATT" (0x0007)
    uint16: 0x0062
    uint16: 0x0065
Language Base Attr List:
  code_ISO639: 0x656e
  encoding:    0x6a
  base_offset: 0x100
  code_ISO639: 0x6672
  encoding:    0x6a
  base_offset: 0x110
  code_ISO639: 0x6465
  encoding:    0x6a
  base_offset: 0x120
  code_ISO639: 0x6a61
  encoding:    0x6a
  base_offset: 0x130

Service Name: AAP Server
Service RecHandle: 0x4f490514
Service Class ID List:
  UUID 128: 74ec2172-0bad-4d01-8f77-997b2be0722a
Protocol Descriptor List:
  "L2CAP" (0x0100)
    PSM: 4097
Language Base Attr List:
  code_ISO639: 0x656e
  encoding:    0x6a
  base_offset: 0x100
  code_ISO639: 0x6672
  encoding:    0x6a
  base_offset: 0x110
  code_ISO639: 0x6465
  encoding:    0x6a
  base_offset: 0x120
  code_ISO639: 0x6a61
  encoding:    0x6a
  base_offset: 0x130
Profile Descriptor List:
  "" (0x4b6f7c74-07f4-49de-b0b9-ab4304728f29)
    Version: 0x0100

Service Name: Handsfree
Service RecHandle: 0x4f49111e
Service Class ID List:
  "Handsfree" (0x111e)
  "Generic Audio" (0x1203)
Protocol Descriptor List:
  "L2CAP" (0x0100)
  "RFCOMM" (0x0003)
    Channel: 7
Language Base Attr List:
  code_ISO639: 0x656e
  encoding:    0x6a
  base_offset: 0x100
  code_ISO639: 0x6672
  encoding:    0x6a
  base_offset: 0x110
  code_ISO639: 0x6465
  encoding:    0x6a
  base_offset: 0x120
  code_ISO639: 0x6a61
  encoding:    0x6a
  base_offset: 0x130
Profile Descriptor List:
  "Handsfree" (0x111e)
    Version: 0x0106

Service Name: AVRCP Target
Service RecHandle: 0x4f49110c
Service Class ID List:
  "AV Remote Target" (0x110c)
Protocol Descriptor List:
  "L2CAP" (0x0100)
    PSM: 23
  "AVCTP" (0x0017)
    uint16: 0x0104
Language Base Attr List:
  code_ISO639: 0x656e
  encoding:    0x6a
  base_offset: 0x100
  code_ISO639: 0x6672
  encoding:    0x6a
  base_offset: 0x110
  code_ISO639: 0x6465
  encoding:    0x6a
  base_offset: 0x120
  code_ISO639: 0x6a61
  encoding:    0x6a
  base_offset: 0x130
Profile Descriptor List:
  "AV Remote" (0x110e)
    Version: 0x0105

Service Name: AVRCP Controller
Service RecHandle: 0x4f49110e
Service Class ID List:
  "AV Remote" (0x110e)
  "AV Remote Controller" (0x110f)
Protocol Descriptor List:
  "L2CAP" (0x0100)
    PSM: 23
  "AVCTP" (0x0017)
    uint16: 0x0104
Language Base Attr List:
  code_ISO639: 0x656e
  encoding:    0x6a
  base_offset: 0x100
  code_ISO639: 0x6672
  encoding:    0x6a
  base_offset: 0x110
  code_ISO639: 0x6465
  encoding:    0x6a
  base_offset: 0x120
  code_ISO639: 0x6a61
  encoding:    0x6a
  base_offset: 0x130
Profile Descriptor List:
  "AV Remote" (0x110e)
    Version: 0x0105

Service Name: Audio Sink
Service RecHandle: 0x4f49110b
Service Class ID List:
  "Audio Sink" (0x110b)
Protocol Descriptor List:
  "L2CAP" (0x0100)
    PSM: 25
  "AVDTP" (0x0019)
    uint16: 0x0103
Language Base Attr List:
  code_ISO639: 0x656e
  encoding:    0x6a
  base_offset: 0x100
  code_ISO639: 0x6672
  encoding:    0x6a
  base_offset: 0x110
  code_ISO639: 0x6465
  encoding:    0x6a
  base_offset: 0x120
  code_ISO639: 0x6a61
  encoding:    0x6a
  base_offset: 0x130
Profile Descriptor List:
  "Advanced Audio" (0x110d)
    Version: 0x0103
```

I get some responses that are expected (as seen on iPad OS 18.2, using PacketLogger), but some are totally different. Also, controlling the case charging sounds (on APP2) uses this.

**Comment by vulpes2** [comment]
Currently trying to fix a really annoying bug in wireshark, it makes tracing captured USB HCI traffic really difficult. Will resume working on the AACP stuff afterwards.

**Comment by kavishdevar** [comment]
The battery status and be received even when the case is not open and the AirPods are not connected. And I don't think any private keys are needed for this. It is just a GATT server that the case has, not the airpods. So I think the socket I was talking about eariler was the case's and not the buds'.

So even when the case is closed, the iPad battery widget can update the charging status, etc.. I believe this is the same thing used to play sound from the case too!!

![image](https://github.com/user-attachments/assets/16f3cfb2-d488-4998-a09b-cf4a7dccc09c)

**Comment by steam3d** [comment]
@kavishdevar I could not find a way to identify that this case is yours.

**Comment by kavishdevar** [comment]
yeah... And I have also noticed the mac address change. But I just had all hy other apple devices turned off and AirPods really really close to my phone while testing it out. I am still trying to decode the accessibility settings (transparency mode customization etc.) for APP2. And then I'll look into getting batter status even when the airpods are disconnected.

**Comment by steam3d** [comment]
@kavishdevar Yes, AirPods also change their MAC address approximately every 16 minutes. AirPods, like the AirPods's case, require a special key to resolve the random MAC address and identify them.

**Comment by vulpes2** [comment]
I'm still busy with some other stuff but you can expect AACP docs by the end of the year. For resolving the RPA you need the IRK, you should be able to find it in your keychain on macOS.

**Comment by steam3d** [comment]
@vulpes2 Yes, but the problem is that usually users don't have a Mac. We need some kind of workaround. Perhaps we can request it through AAP.

**Comment by vulpes2** [comment]
Indeed, I'll investigate and report back when I figure it out.

**Comment by kavishdevar** [comment]
@vulpes2, have you tried decoding the head tracking packets? I've spent quite some time trying to do that to implement head gesures on my android app, but I can't figure out anything. I've tried reading 4 bytes as one value, signed integers, but have had no luck.

Please tell if you can understand the format of the packet sent.

**Comment by steam3d** [comment]
@kavishdevar Send us an example packet so we don't waste time searching

**Comment by steam3d** [comment]
@kavishdevar Got it, I thought you were trying to set up new head gestures for answering calls. Okay, when I have time I'll see how it works on my headphones. Although it seems to me that it would be strange to transfer data about the position of the head to the iPhone. It is much easier to process sound and gestures on the AirPods themselves, since this data would be quite heavily loaded on the communication channel.I would guess that the headphones and iPhone exchange each other's position from time to time, just to sync which side the iPhone is facing. But who knows how Apple implemented it.

**Comment by kavishdevar** [comment]
> Although it seems to me that it would be strange to transfer data about the position of the head to the iPhone. It is much easier to process sound and gestures on the AirPods themselves, since this data would be quite heavily loaded on the communication channel.
I wish that was true, wouldn't have had to decode the whole packet format...

But, I used PacketLogger on my macbook (non-Apple Silicon), I found out that whenever Siri wants confirmation, it continuously sends these packets. When I plotted the live packets from my linux machine, it was cleary visible that movements are sent over the socket!

> I would guess that the headphones and iPhone exchange each other's position from time to time, just to sync which side the iPhone is facing. But who knows how Apple implemented it.

I think this might be true for spatial audio, but the last time I tried it (M1 MacBook Air), the position of the laptop didn't affect the audio. The change was clear only when I moved myself, and not when I moved the laptop.

> Okay, when I have time I'll see how it works on my headphones.
Thank you! :)

**Comment by kavishdevar** [comment]
Also, about the Bluetooth GATT server running on the case - I read a Find My Network Accessory Specification document (fairly recent, around 3-4 years old), and found out that this is where the play sound is supposed to go. Also, the accurate battery percentages are also a characteristic on this GATT server, on what Apple calls a "Owner Control Point".

Retreiving the IRK is not the only challenge now, the iphone/ipad must authenticate as a "Owner" with the accessory, in order to get battery information, play sound, etc.. vague battery levels like high, low, full, are visible in the nrf connect app, which makes me believe that this is the only thing that can be achieved by just the IRK.

Interestingly, there is a non-owner control point too! which i think is to be able to play sound when the device is being unwantedly tracked. (from the spec. I found)

So, it might be possible to play a sound without "authenticating as owner", but I don't have a iPhone to trigger an unwanted tracking detection.

**Comment by kavishdevar** [comment]
> in order to get battery information

Basically, the battery widgets have accurate levels even when they're not connected (the case is closed, and the pods are in case.).

**Comment by steam3d** [comment]
> Also, about the Bluetooth GATT server running on the case - I read a Find My Network Accessory Specification document (fairly recent, around 3-4 years old), and found out that this is where the play sound is supposed to go. Also, the accurate battery percentages are also a characteristic on this GATT server, on what Apple calls a "Owner Control Point".

It seemed to me that Apple was using a hidden BLE GATT and just connecting to it to get the charge level. Since there is no particular point in complicating everything so much, since essentially the charge level values ​​​​is read only

**Comment by kavishdevar** [comment]
Not just for the charging levels, but also writing other settings that need to go to the case, like case charging sounds, playing sounds, etc... Maybe the accurate battery status when not connected are available in non-owner control point, I'll have a look later once I'm done with head gestures.

**Comment by vulpes2** [comment]
I have acquired some more hardware recently, but haven't had time to look into the head gesture stuff yet. I can poke at it later this week, is this data sent over AACP or GATT?

**Comment by kavishdevar** [comment]
Head tracking over AACP (check out my python script if you would like..)

**Comment by vulpes2** [comment]
I'm back to working on this again, trying to write a proper decoder for all the stuff that I have figured out so far. As for the head gestures, I don't think the data you had was related to them, because I have observed those from the first gen AirPods Pros too, which definitely does not support that feature. I think it's more likely related to the spatial audio stuff instead.

**Comment by steam3d** [comment]
@vulpes2 as @kavishdevar mentions Siri, I suppose it could be data from the microphone, recognized text from voice input, or something similar.

**Comment by kavishdevar** [comment]
Nop, it is not, I typed to siri to make sure, and A2DP was hidden using the priority order @steam3d.
And, @vulpes2, These packetes are head tracking. They're formatted the same for head gestures, and for spatial audio. It is just software that doesn't allow the Pro 1 to use head gestures.

**Comment by kavishdevar** [comment]
I tried Spatial Audio on an M1 MBA and it was the same packetes as the head gestures!

**Comment by vulpes2** [comment]
It's definitely not voice data as those would be sent over SCO, at least on Sequoia there doesn't seem to be a way to log them without sniffing the PCIe traffic directly (Apple has been doing HCI over PCIe for quite a long time now).

**Comment by kavishdevar** [comment]
Install a Bluetooth bug reporting profile for Mac and use PacketLogger from Additional Tools for XCode @vulpes2

**Comment by vulpes2** [comment]
I'm talking about SCO packets, not ACL packets, SCO packets do not get logged even with the profile installed.

**Comment by vulpes2** [comment]
Unfortunately the AAP/AACP that we are working with is completely different to the old Apple Accessory Protocol that the iPod used back in the day. Had to put this project on hold because life got in the way, will get back to this in a bit. @kavishdevar Have you been able to get any BT packet captures on an iOS device? Even with the correct logging profile installed I'm still not getting anything on iOS 18.

**Comment by kavishdevar** [comment]
Yeah, I have been able to capture with the profile installed, albiet on iPadOS, 18.3. It did stop working for a while on 18.2. 

Also, I have worked out the “Customize transparency mode” packet format, but for some reason it doesn't have any effect when sent from Linux/Android. I tried sending the exact same packet macOS sends instead creating my own, but no luck. Most likely, there is another packet that needs to be sent before this, just like conversational awareness and adaptive transparency needs. And, seems like Apple hides those.

Since Apple hides a few packets, I was wondering if I can get some hardware and sniff directly. I know I can disable the internal wireless module and force macOS to use a dongle. I even have a hackintosh, with SIP disabled with which I use my airpods often. There might be a way to directly capture the packets without even needing some special hardware, just maybe!

**Comment by kavishdevar** [comment]
Also, would our work have been easier if it was the same? I never thought there were any similarities in the first place. @vulpes2

**Comment by vulpes2** [comment]
Good to know it's finally fixed, it was broken on 18.0 - 18.2 so I didn't bother testing 18.3.

Sniffing hardware traffic is possible but you will need a pretty old Mac (or your hackintosh), the Bluetooth module on them is just a regular USB device that you can capture traffic within macOS without additional external hardware. On a hackintosh try https://github.com/OpenIntelWireless/IntelBluetoothFirmware on an AX200, you don't need an Apple branded Broadcom card for the AACP and A2DP parts. WBS on HFP won't work on any of the USB chipsets (Broadcom included) but it's not relevant for now.

I've had a ton of problems with Wireshark's USB dissectors. While it will give you hexdumps, the `bthci` layer will not dissect correctly due to issues spanning across a few different dissectors. Spent some time debugging it last year, but didn't have enough time to actually fix anything. This is something I'm planning to look into again when I have more time later this year.

**Comment by kavishdevar** [comment]
Has someone been able to identify AirPods with BLE adverts with the IRK? I was trying the standard algorithm for RPAs, but AirPods don't even advertize with a RPA, it seems like a non-resolvable randomized address. Also, the IRK can definitely be fetched from AirPods with AAP without requiring a Mac. I just can't figure out how to use it.

**Comment by steam3d** [comment]
@kavishdevar  If you change the BLE random MAC to that of real AirPods, the iPhone will detect them as your AirPods, but if you set the BLE MAC to a manually chosen random value, the iPhone will not detect them. I suppose the iPhone can resolve the MAC address.

**Comment by kavishdevar** [comment]
That's not what I meant, @steam3d. I meant that I can't figure out how to use the IRK that I have from my macOS Keychain. I am assuming this IRK is supposed to help identify the AirPods' BLE broadcasts used by apps like CAPod, but I can't figure out how to do that.

@vulpes2 mentioned [here](https://github.com/steam3d/MagicPodsCore/issues/14#issuecomment-2482661711) that we should be able to resolve the RPA. But I don't think the address used by AirPods is even a RPA. It seems to me like a random random address.

**Comment by steam3d** [comment]
@kavishdevar If that were the case, the iPhone would trigger the popup animation using any MAC, but it will not do anything until it recognizes the known random MAC. Above, I described a method to test it.

AirPods change their MAC every 15 minutes, so you have 15 minutes to copy the BLE packet and emulate it on the "hacker device." The iPhone will detect the "hacker device" as AirPods because the random MAC and packet data are valid for the iPhone. If you change the random MAC to your own value, the iPhone will not detect the AirPods and will not display the popup animation.

This means that the iPhone can resolve this MAC and determine whether these are known AirPods or not. You can deceive the iPhone by sending the real MAC of the AirPods and the same data as before. This means that the AirPods resolve the MAC address to the real MAC address and compare the values. If the resolved MAC equals the real MAC, then these are known AirPods.

Why not the data? Because if the iPhone were detecting based on BLE data, it would display the popup animation with any MAC, but it does not do that.

**Comment by kvtbuyharita-png** [comment]
> [@kavishdevar](https://github.com/kavishdevar)ฉันไม่สามารถหาวิธีระบุได้ว่าคดีนี้เป็นของคุณ

[org.fdroid.fdroid-20260420T170515Z.txt](https://github.com/user-attachments/files/26905082/org.fdroid.fdroid-20260420T170515Z.txt)

**Comment by kvtbuyharita-png** [comment]
> [@kavishdevar](https://github.com/kavishdevar)ฉันไม่สามารถหาวิธีระบุได้ว่าคดีนี้เป็นของคุณ

[org.fdroid.fdroid-20260420T170515Z.txt](https://github.com/user-attachments/files/26905082/org.fdroid.fdroid-20260420T170515Z.txt)

**Comment by kvtbuyharita-png** [comment]
> [@kavishdevar](https://github.com/kavishdevar)ถ้าเป็นเช่นนั้น iPhone จะแสดงภาพเคลื่อนไหวป๊อปอัพโดยใช้ MAC ใดก็ได้ แต่จะไม่ทำงานใดๆ จนกว่าจะรู้จัก MAC แบบสุ่มที่ทราบแล้ว ข้างต้นผมได้อธิบายวิธีการทดสอบไว้แล้ว
> 
> AirPods จะเปลี่ยนที่อยู่ MAC ทุก 15 นาที ดังนั้นคุณมีเวลา 15 นาทีในการคัดลอกแพ็กเก็ต BLE และจำลองมันบน "อุปกรณ์แฮ็กเกอร์" iPhone จะตรวจจับ "อุปกรณ์แฮ็กเกอร์" ว่าเป็น AirPods เพราะที่อยู่ MAC แบบสุ่มและข้อมูลแพ็กเก็ตนั้นใช้ได้กับ iPhone หากคุณเปลี่ยนที่อยู่ MAC แบบสุ่มเป็นค่าของคุณเอง iPhone จะไม่ตรวจจับ AirPods และจะไม่แสดงภาพเคลื่อนไหวป๊อปอัพ
> 
> หมายความว่า iPhone สามารถตรวจสอบ MAC address นี้และระบุได้ว่านี่คือ AirPods ที่รู้จักหรือไม่ คุณสามารถหลอก iPhone ได้โดยการส่ง MAC address จริงของ AirPods และข้อมูลเดียวกันกับที่ใช้ก่อนหน้านี้ ซึ่งหมายความว่า AirPods จะตรวจสอบ MAC address เพื่อหา MAC address จริงและเปรียบเทียบค่า หาก MAC address ที่ตรวจสอบแล้วเท่ากับ MAC address จริง แสดงว่าเป็น AirPods ที่รู้จัก
> 
> ทำไมถึงไม่ใช้ข้อมูล? เพราะถ้า iPhone ตรวจจับโดยใช้ข้อมูล BLE มันควรจะแสดงภาพเคลื่อนไหวป๊อปอัพกับ MAC ทุกเครื่อง แต่ iPhone กลับไม่ทำอย่างนั้น
https://github.com/steam3d/MagicPodsCore/issues/14#issuecomment-2481420153

**Comment by kvtbuyharita-png** [comment]
> Apple ได้เปลี่ยนชื่อโปรโตคอลเป็น AACP แล้ว แต่ยังไม่ทราบแน่ชัดว่าย่อมาจากอะไร อาจจะเป็น Apple Audio Configuration Protocol หรือเปล่า? อย่างไรก็ตาม นี่คือข้อมูลโครงสร้างพื้นฐานสำหรับข้อความ AACP
> 
> แพ็กเก็ต AACP ประกอบด้วยฟิลด์ประเภท u16 ฟิลด์รหัสบริการ u16 และเพย์โหลด รหัสบริการจะเป็น 4 เสมอหลังจากสร้างการเชื่อมต่อแล้ว ยังไม่พบรหัสบริการอื่น ๆ ในขณะนี้
> 
> มีแพ็กเก็ตอยู่ห้าประเภท:
> 
> * 0: เชื่อมต่อ, ข้อมูลที่ส่ง:`u16 major, u16 minor, u64 features`
> * 1: การตอบสนองการเชื่อมต่อ, เพย์โหลด:`u16 status, u16 major, u16 minor, u64 features`
> * 2: ตัดการเชื่อมต่อ, ข้อมูลที่ส่ง:`u16 status`
> * 3: การตอบสนองการตัดการเชื่อมต่อ ไม่มีข้อมูลส่ง
> * 4: ข้อความ, เนื้อหา:`u16 opcode, payload`
> 
> ขนาดของข้อมูลในแพ็กเก็ตข้อความ AACP จะแตกต่างกันไปตามโอเปอเรชันโค้ด โดยอาจมีขนาดคงที่หรือมีฟิลด์ระบุขนาดอยู่ ตัวอย่างเช่น โอเปอเรชันโค้ด 0x9 คือข้อความสำหรับการกำหนดค่าที่มีขนาดข้อมูลคงที่ 5 ไบต์ ไบต์แรกคือประเภทของการกำหนดค่า และ 4 ไบต์ถัดไปคือค่าต่างๆ ไม่ใช่ทุกการตั้งค่าที่จะใช้ทั้ง 4 ไบต์ ส่วนใหญ่จะเว้นไบต์ที่ไม่ได้ใช้ไว้เป็น 0x00
> 
> เอกสารเพิ่มเติมจะตามมาในอีกไม่กี่สัปดาห์ข้างหน้า แต่ตอนนี้ฉันยังไม่มีเวลาเขียน ดูเหมือนว่า Handoff จะเกี่ยวข้องกับ iCloud และอาจเกินความสามารถของฉันที่จะแก้ไขได้ ความพยายามส่วนใหญ่ของฉันจนถึงตอนนี้มุ่งเน้นไปที่การวิเคราะห์ข้อความ AACP ทั้งหมดที่ฉันรวบรวมได้

https://github.com/user-attachments/files/26905082/org.fdroid.fdroid-20260420T170515Z.txt
